### PR TITLE
Preinstalled Java

### PR DESCRIPTION
**Pull request by rpelisse**
This PR contains a molecule scenario to reproduce the issue 206, where the target system has already a preinstalled JDK. It also contains the fix for it.

### CODE REVIEW

**Comment by guidograzioli** [comment]
Change is good for me; you only need to add `jws_java_home` in argument_specs with the same options as `jws_java_version`.
@defeatedbylife the behaviour you request is implemented with the fallback here: https://github.com/ansible-middleware/jws/blob/68f6da06dc2cc510379238f9725fab07746fe245/roles/jws/tasks/systemd/systemd.yml#L54

`/usr/lib/jvm/java` is the final JAVA_HOME path when neither `jws_java_version` or `jws_java_home` is set.

**Comment by rpelisse** [comment]
>From what I can see, this doesn't address the problem that if I don't set the jws_java_home, the default installed JDK won't be picked up

Just to be clear, that was never a use case we meant to support (at least for this version). Either the user let the collection install the required dependency (the jdk the server needs), either the user provide the JAVA_HOME of the already installed JDK (that was broken and is now fixed by this PR).

> I wonder if we can limit the setting up JAVA_HOME to cases when it is set through the collection either manually through jws_java_home or jws_rpm_java_home?

That can be easily implemented indeed and it would mean that the server starts with whatever java version is already installed (if any) on the target system:
```
{%if jws_java_home is defined or jws_rpm_java_home is defined %}
# Where your java installation lives
JAVA_HOME={{ jws_java_home  | default(jws_rpm_java_home, true) }}
{%endif%}
```
Now, I'm not sure if it's better to let the server starts with an undefined Java version. It seems to me that an admin system (or a devops specialist) would prefer to have something more strict. To be clear, I don't mind it, it's just an extra if stmt in the template.

**Comment by guidograzioli** [comment]
0.02$ here: maybe we can render the line more human readable by doing:

```
JAVA_HOME={{ jws_java_home  | default(jws_rpm_java_home, true) | default('/usr/lib/jvm/java') }}
```
which fully describe the "Use the user-specified jdk installation if defined, or use the user-specified jdk rpm id defined, or finally use the default  /usr/lib/jvm/java"

**Comment by rpelisse** [comment]
@guidograzioli added the "human friendly" version, plus the missing entry in argument_specs.

**Comment by rpelisse** [comment]
@defeatedbylife oh good remarks, let me add those right away.

**Comment by guidograzioli** [comment]
> @guidograzioli However, does `/usr/lib/jvm/java` exists? I would say not always. And certainly not in the containers that are running during my tests.

Which rhel version? The symlink should exists as soon any JDK rpm is installed (via the alternatives config). If no JDK is installed via rpm, and no JDK is installed manually (hence with configured path in the playbook), a failure is to be expected here

**Comment by guidograzioli** [comment]
So if we can't rely on the symlink to be there, our only way is to make `jws_java_home` XOR `jws_java_version` required in a prerequisite check, @rpelisse what do you reckon?

**Comment by rpelisse** [comment]
Weird (about the missing link), but yes, I'll retcon that and remove the symlink. This is also showing why we can't be too smart here, so much things may change depending on the setup.

**Comment by rpelisse** [comment]
@guidograzioli I reworked the molecule scenario to use a downloaded JDK (to keep the test a bit more honest). Can you take one last look and approve? (don't merge for now).

File: roles/jws/meta/argument_specs.yml [hunk 1]
```diff
@@ -357,6 +357,9 @@ argument_specs:
                 default: "{{ lookup('env', 'PWD') | default('/opt') }}"
                 description: "path local to controller for offline/download archive files"
                 type: "str"
+            jws_java_home:
+                description: "path to the JAVA_HOME to be used by the server"
+                type: "str"
```
Comment (guidograzioli) [inline roles/jws/meta/argument_specs.yml:362]: you need `required: false` here to have the variable not always defined (but then the README section above should be updated to reflect that); or, `default: ""` to have it always defined as an empty string (if not overridden)

**Review by guidograzioli — CHANGES_REQUESTED** [review]

File: roles/jws/meta/argument_specs.yml (hunk 1 above)
Comment (rpelisse) [reply roles/jws/meta/argument_specs.yml:363]: Thanks! Fixed!

**Review by guidograzioli — APPROVED** [review]

**Pull request merged**